### ISSUE TITLE
feat(nn): FiLM conditional modulation layer for neural fields

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#86]
## Summary

Add a Feature-wise Linear Modulation (FiLM) layer to `pyrox.nn` so users can build *conditional* neural fields — a single model `f(x; z)` whose behaviour at coordinate `x` is modulated by a latent / side-information code `z`. FiLM is the workhorse conditioning primitive behind conditional NeRFs, latent SIREN, and conditional Bayesian neural fields; it slots into the same `PyroxModule` pattern as `DenseReparameterization`, `VariationalFourierFeatures`, etc., and is the missing piece for BNF-style ensembles that share structure across series / locations.

Landing this enables three follow-up directions without new infrastructure: (a) conditional BNF (one model, many series), (b) hypernetwork-style amortized GPs where `z` encodes a task, (c) latent-code MAP/VI where the code itself becomes an inferred latent.

## Background

FiLM (Perez et al., AAAI 2018) is the simplest and most empirically robust way to condition a network on side information. Given an intermediate activation `h ∈ R^C` and a conditioning vector `z ∈ R^K`, FiLM applies a feature-wise affine transform whose parameters are produced from `z`:

$$
\mathrm{FiLM}(h; z) = \gamma(z) \odot h + \beta(z),
\qquad \gamma, \beta : \mathbb{R}^{K} \to \mathbb{R}^{C}.
$$

Why this matters for pyrox:

- It's the standard adapter between a **coordinate network** (SIREN, HSGP features, VFF) and a **latent code** — the piece missing today.
- In Bayesian mode (priors on the generator of γ, β), a single FiLM block makes `BayesianNeuralField` conditional with no architecture rewrite.
- It commutes with every existing `PyroxModule` — FiLM is just `y = γ ⊙ h + β`, so it can wrap any layer output.

### References

- Perez, Strub, de Vries, Dumoulin, Courville. **"FiLM: Visual Reasoning with Conditional Feature-wise Linear Modulation."** AAAI 2018. [arXiv:1709.07871](https://arxiv.org/abs/1709.07871)
- Dumoulin et al. **"Feature-wise transformations."** Distill 2018 — survey of conditioning styles (AdaIN, FiLM, CBN). [distill.pub/2018/feature-wise-transformations](https://distill.pub/2018/feature-wise-transformations/)
- Mehta et al. **"Modulated Periodic Activations for Generalizable Local Functional Representations."** ICCV 2021 — FiLM-style modulation over SIREN. [arXiv:2104.03960](https://arxiv.org/abs/2104.03960)
- Chan et al. **"π-GAN."** CVPR 2021 — uses FiLM on SIREN for conditional neural fields.

## Proposed API

New public layers in `src/pyrox/nn/_layers.py`, re-exported from `pyrox.nn`.

### Core deterministic layer

```python
class FiLM(PyroxModule):
    """Feature-wise Linear Modulation: ``y = γ(z) ⊙ h + β(z)``.

    Produces ``(γ, β)`` from the conditioning vector ``z`` via a single
    ``eqx.nn.Linear`` generator of width ``2 * num_features``. The two
    halves are split on the output axis; ``γ`` is passed through a
    user-selected transform (identity by default; ``1 + tanh`` for
    stable-init) so that at ``z = 0`` the layer is close to identity.
    """
    generator: eqx.nn.Linear        # R^K -> R^{2C}
    num_features: int               # C
    cond_dim: int                   # K
    gamma_activation: str = "identity"   # "identity" | "one_plus_tanh" | "softplus"

    @classmethod
    def init(
        cls,
        num_features: int,
        cond_dim: int,
        *,
        key: PRNGKeyArray,
        gamma_activation: str = "one_plus_tanh",
        pyrox_name: str | None = None,
    ) -> "FiLM": ...

    @pyrox_method
    def __call__(
        self,
        h: Float[Array, "N C"],
        z: Float[Array, "N K"] | Float[Array, " K"],
    ) -> Float[Array, "N C"]: ...
```

Semantics:

- `h` shape `(N, C)` or `(C,)`; `z` shape `(N, K)` (per-sample) or `(K,)` (broadcast to all rows).
- Generator is a single affine map — not a deep MLP — to match the minimalism of `DenseReparameterization` and to keep the prior count tractable when we wrap it in the Bayesian version. If users want a deep generator they can compose FiLM(`g(z)`) themselves.
- `gamma_activation` default is `"one_plus_tanh"` so γ starts near 1 and β near 0 → layer is near-identity at init, matching the established practice in Perez et al.

### Bayesian variant

```python
class BayesianFiLM(PyroxModule):
    """FiLM with priors on the generator weights.

    Registers NumPyro sites ``{scope}.gen_W`` and ``{scope}.gen_b``
    (via ``pyrox_sample``) under the user's chosen prior. Defaults
    to ``Normal(0, prior_std)`` on both, which recovers the Bayesian
    linear regression prior in the γ/β channel.
    """
    prior_std: float = 1.0
    # + same public surface as FiLM
```

Registers two sites — nothing fancier. Users who want Flipout / Reparameterization in the generator can do `DenseReparameterization` composed with `FiLM(... gamma_activation=...)` manually.

### Convenience helper

```python
def apply_film(
    h: Float[Array, "N C"],
    gamma: Float[Array, "N C"] | Float[Array, " C"],
    beta: Float[Array, "N C"] | Float[Array, " C"],
) -> Float[Array, "N C"]:
    """Pure-JAX FiLM op for users who compute ``γ, β`` externally."""
```

This is the escape hatch for hypernetwork-style use cases where the user already has `(γ, β)` from a bespoke generator (e.g., an amortized encoder over a graph of series).

## Integration points

- **SIREN / INR**: FiLM after each Dense sine layer → modulated SIREN (Mehta et al. 2021). Unblocks the planned SIREN layer issue.
- **`BayesianNeuralField`** (the bayesnf port): wrapping each MLP block's output with `BayesianFiLM(cond_dim=K)` turns the field into a conditional field over a per-series latent `z`. Zero math change to the existing BNF likelihood code.
- **`HSGPFeatures` / `VariationalFourierFeatures`**: applying FiLM on the spectral feature output `Φ(x)` before the linear readout amounts to conditioning the *amplitude* and *offset* of each spectral component — effectively a conditional spectral density. Interesting ablation.
- **Ensemble-of-MAP runner** (planned in the bayesnf port): the ensemble axis is naturally a "conditioning draw" — a FiLM block fed a per-ensemble-member code reproduces the amortized-ensemble trick from Izmailov et al. 2021 without writing new inference code.

### File layout

- **Edit** `src/pyrox/nn/_layers.py`: add `FiLM`, `BayesianFiLM`, `apply_film`. Keep parity with the existing 12 layers (PyroxModule subclass, `.init` classmethod, `@pyrox_method __call__`, scope-aware sample sites).
- **Edit** `src/pyrox/nn/__init__.py`: re-export the three new public names.
- **New** `tests/nn/test_film.py`: unit + property tests (see below).
- **New** `docs/api/nn/film.md`: mkdocstrings stubs — three `::: pyrox.nn.FiLM` blocks.
- **Edit** `mkdocs.yml`: nav entry under `nn/`.

### Math/init details (load-bearing)

- Generator output split: use `einops.rearrange(out, "... (two c) -> two ... c", two=2)` to avoid raw `jnp.split`. Follows gaussx convention D9 (einops for all reshape/einsum).
- `gamma_activation="one_plus_tanh"`: `γ = 1 + jnp.tanh(raw_gamma)`; β stays raw. This gives `E[γ]=1, E[β]=0` under zero-mean generator weights → identity at init, which is the prior we want for "FiLM does nothing until trained" behaviour.
- `z` broadcast rule: if `z.ndim == 1`, treat as a single per-batch code and broadcast to `(N, K)` via `einops.repeat`. Document this explicitly — it's a common source of shape bugs.

### Handling `h.ndim == 1`

`h` of shape `(C,)` → treat as a single-row activation; squeeze result back to `(C,)`. Implement via `jnp.atleast_2d` on both `h` and `z`, run the 2D path, then conditionally squeeze — documented in the docstring with a shape table.

### Type annotations

```python
from jaxtyping import Array, Float, PRNGKeyArray
```

Per-file ruff ignore already in effect via `"src/pyrox/nn/**" = ["F722"]` (jaxtyping string annotations). No new config.

## Tests — `tests/nn/test_film.py`

All tests use the existing numpyro/equinox fixtures from the current `tests/nn/test_spectral_layers.py` for consistency.

1. **`test_film_output_shape`** — `FiLM(num_features=8, cond_dim=4)` on `h: (6, 8)`, `z: (6, 4)` → output `(6, 8)`.
2. **`test_film_broadcasts_scalar_conditioning`** — `z: (4,)` broadcasts over all `N`; compare against an explicit `einops.repeat(z, "k -> n k", n=6)` baseline.
3. **`test_film_identity_at_init_with_one_plus_tanh`** — with `gamma_activation="one_plus_tanh"`, at init (generator bias=0), output ≈ input to `atol=1e-5`. This is the load-bearing property that justifies the default.
4. **`test_film_registers_generator_sites`** — `BayesianFiLM` traced under `handlers.trace()` registers exactly `{scope}.gen_W` and `{scope}.gen_b` sample sites.
5. **`test_film_jits`** — forward pass under `jax.jit` returns finite values. Mirrors the existing `test_hsgp_jits` pattern.
6. **`test_apply_film_matches_module`** — the pure helper `apply_film(h, γ, β)` matches `FiLM.__call__` when given the same `(γ, β)`, verifying the escape hatch is consistent.
7. **`test_film_composes_with_siren`** *(conditional-skip if SIREN not yet landed)* — SIREN sine-Dense → FiLM → SIREN sine-Dense produces finite outputs and non-trivial gradient on `z`. Integration smoke test.
8. **`test_film_rejects_mismatched_dims`** — `ValueError` on `z.shape[-1] != cond_dim` and on `h.shape[-1] != num_features`, with messages containing `"cond_dim"` and `"num_features"` respectively.

Keep tests fast — no training loops. Direct shape / value assertions only.

## Docs

- mkdocstrings page `docs/api/nn/film.md` with three `:::` blocks.
- A 4-6 line usage snippet in the page body showing the `SIREN → FiLM → SIREN` composition (feel free to mark TODO until SIREN lands).
- **No notebook** in this issue — add a FiLM section to the planned `bayesian_neural_fields.py` notebook once conditional BNF lands.

## Acceptance criteria

- [ ] `pyrox.nn.FiLM`, `pyrox.nn.BayesianFiLM`, `pyrox.nn.apply_film` are public and documented.
- [ ] 8 tests in `tests/nn/test_film.py` all pass locally and under CI.
- [ ] `uv run pytest -q` green.
- [ ] `uv run --group lint ruff check .` green (entire repo).
- [ ] `uv run --group lint ruff format --check .` green (entire repo).
- [ ] `uv run --group typecheck ty check src/pyrox` green.
- [ ] Identity-at-init property test passes to `atol=1e-5` — this is the non-negotiable correctness check.

## Out of scope (explicit)

- Deep generator MLPs — users compose themselves.
- AdaIN / CBN variants — FiLM is the only conditioning primitive in this issue.
- SIREN integration (separate issue).
- Conditional BNF estimator (separate issue, depends on this + SIREN + the planned bayesnf port).
- Hypernetworks where weights themselves (not γ/β) are generated from `z` — different design, different issue.

## Estimated effort

Small. `FiLM` itself is ~30 lines; `BayesianFiLM` is another ~20 lines; tests and docs ~150 lines total. Single PR, no new dependencies.
